common/render_shaders.go

```diff
@@ -783,6 +775,320 @@ func (l *legacyShader) Post() {
+	l.projectionMatrix[8] = 1
+
+	l.viewMatrix = make([]float32, 9)
+	l.viewMatrix[0] = 1
+	l.viewMatrix[4] = 1
+	l.viewMatrix[8] = 1
+
+	l.modelMatrix = make([]float32, 9)
+	l.modelMatrix[0] = 1
+	l.modelMatrix[4] = 1
+	l.modelMatrix[8] = 1
+
+	if l.cameraEnabled {
+		for _, system := range w.Systems() {
+			switch sys := system.(type) {
+			case *CameraSystem:
+				l.camera = sys
+			}
+		}
+		if l.camera == nil {
+			log.Println("WARNING: BasicShader has CameraEnabled, but CameraSystem was not found")
+		}
+	}
+
+	return nil
+}
+
+func (l *textShader) Pre() {
+	engo.Gl.Enable(engo.Gl.BLEND)
+	engo.Gl.BlendFunc(engo.Gl.SRC_ALPHA, engo.Gl.ONE_MINUS_SRC_ALPHA)
+
+	// Bind shader and buffer, enable attributes
+	engo.Gl.UseProgram(l.program)
+	engo.Gl.BindBuffer(engo.Gl.ELEMENT_ARRAY_BUFFER, l.indicesRectanglesVBO)
+	engo.Gl.EnableVertexAttribArray(l.inPosition)
+	engo.Gl.EnableVertexAttribArray(l.inTexCoords)
+	engo.Gl.EnableVertexAttribArray(l.inColor)
+
+	if engo.ScaleOnResize() {
+		l.projectionMatrix[0] = 1 / (engo.GameWidth() / 2)
+		l.projectionMatrix[4] = 1 / (-engo.GameHeight() / 2)
+	} else {
+		l.projectionMatrix[0] = 1 / (engo.CanvasWidth() / 2)
+		l.projectionMatrix[4] = 1 / (-engo.CanvasHeight() / 2)
+	}
+
+	if l.cameraEnabled {
+		l.viewMatrix[1], l.viewMatrix[0] = math.Sincos(l.camera.angle * math.Pi / 180)
+		l.viewMatrix[3] = -l.viewMatrix[1]
+		l.viewMatrix[4] = l.viewMatrix[0]
+		l.viewMatrix[6] = -l.camera.x
+		l.viewMatrix[7] = -l.camera.y
+		l.viewMatrix[8] = l.camera.z
+	} else {
+		l.viewMatrix[6] = -1 / l.projectionMatrix[0]
+		l.viewMatrix[7] = 1 / l.projectionMatrix[4]
+	}
+
+	engo.Gl.UniformMatrix3fv(l.matrixProjection, false, l.projectionMatrix)
+	engo.Gl.UniformMatrix3fv(l.matrixView, false, l.viewMatrix)
+}
+
+func (l *textShader) updateBuffer(ren *RenderComponent, space *SpaceComponent) {
+	txt, ok := ren.Drawable.(Text)
+	if !ok {
+		unsupportedType(ren.Drawable)
+		return
+	}
+
+	if len(ren.bufferContent) < 20*len(txt.Text) {
+		ren.bufferContent = make([]float32, 20*len(txt.Text)) // TODO: update this to actual value?
+	}
+	if changed := l.generateBufferContent(ren, space, ren.bufferContent); !changed {
+		return
+	}
+
+	if ren.buffer == nil {
+		ren.buffer = engo.Gl.CreateBuffer()
+	}
+	engo.Gl.BindBuffer(engo.Gl.ARRAY_BUFFER, ren.buffer)
+	engo.Gl.BufferData(engo.Gl.ARRAY_BUFFER, ren.bufferContent, engo.Gl.STATIC_DRAW)
+}
+
+func (l *textShader) generateBufferContent(ren *RenderComponent, space *SpaceComponent, buffer []float32) bool {
+	var changed bool
+
+	tint := colorToFloat32(ren.Color)
+	txt, ok := ren.Drawable.(Text)
+	if !ok {
+		unsupportedType(ren.Drawable)
+		return false
+	}
+
+	atlas, ok := atlasCache[txt.Font.URL]
+	if !ok {
+		// Generate texture first
+		atlas = txt.Font.generateFontAtlas(UnicodeCap)
+		atlasCache[txt.Font.URL] = atlas
+	}
+
+	var currentX float32
+	var currentY float32
+
+	var modifier float32 = 1
+	if txt.RightToLeft {
+		modifier = -1
+	}
+
+	for index, char := range txt.Text {
```

Comment:
TODO: we could cache here (for this specific string), so we don't have to compute this again.